src/main/java/io/projectriff/invoker/JavaFunctionInvokerServer.java

```diff
@@ -66,33 +63,24 @@ public JavaFunctionInvokerServer(Function<Flux<?>, Flux<?>> function, Gson mappe
 
 	private Function<Flux<Message<?>>, Flux<Message<?>>> enhance(
 			Function<Flux<Message<?>>, Flux<Message<?>>> function) {
-		return messages -> function.apply(messages.flatMap(m -> push(m)))
-				.flatMap(m -> pop(m)).subscriberContext(setup());
-	}
-
-	private Function<Context, Context> setup() {
-		return ctx -> ctx.put(CONTEXT, new HashMap<String, Object>());
+		AtomicReference<Map<String, Object>> headers = new AtomicReference<>();
+		return messages -> function.apply(messages.map(m -> push(headers, m)))
+				.map(m -> pop(headers, m));
 	}
 
-	private Mono<Message<?>> pop(Message<?> message) {
-		return Mono.subscriberContext().map(ctx -> {
-			Map<String, Object> map = ctx.get(CONTEXT);
-			if (map.isEmpty()) {
-				return message;
-			}
-			Message<?> result = MessageBuilder.fromMessage(message)
-					.copyHeadersIfAbsent(map).build();
-			map.clear();
-			return result;
-		});
+	private Message<?> pop(AtomicReference<Map<String, Object>> headers,
+			Message<?> message) {
+		Map<String, Object> map = headers.getAndSet(Collections.emptyMap());
```

Comment:
@dsyer Why not simply use `get` as the headers value is about to become garbage?

Reply:
Is it? I think it hangs around until the whole stream has been processed (so if you don't `getAndSet()` you can get multiple response messages per request, and the gateway gets confused).

Reply:
Oh you're right. Thanks.